ethcore/private-tx/src/log.rs

```diff
@@ -0,0 +1,409 @@
+		let logging = Logging {
+			logs: RwLock::new(HashMap::new()),
+			logs_serializer,
+			mono_time: Mutex::default(),
+		};
+		match logging.read_logs() {
+			// Initialize time source by max from current system time and max creation time from already saved logs
+			Ok(initial_time) => *logging.mono_time.lock() = MonoTime::new(initial_time),
+			Err(err) => warn!(target: "privatetx", "Cannot read logs: {:?}", err),
+		}
+		logging
+	}
+
+	/// Retrieves log for the corresponding tx hash
+	pub fn tx_log(&self, tx_hash: &H256) -> Option<TransactionLog> {
+		self.logs.read().get(&tx_hash).cloned()
+	}
+
+	/// Logs the creation of the private transaction
+	pub fn private_tx_created(&self, tx_hash: &H256, validators: &[Address]) {
+		let mut validator_logs = Vec::new();
+		for account in validators {
+			validator_logs.push(ValidatorLog {
+				account: *account,
+				validation_timestamp: None,
+			});
+		}
+		let mut logs = self.logs.write();
+		if logs.len() > MAX_JOURNAL_LEN {
+			// Remove the oldest log
+			if let Some(tx_hash) = logs.values()
+				.min_by(|x, y| x.creation_timestamp.cmp(&y.creation_timestamp))
+				.map(|oldest| oldest.tx_hash)
+			{
+				logs.remove(&tx_hash);
+			}
+		}
+		logs.insert(*tx_hash, TransactionLog {
+			tx_hash: *tx_hash,
+			status: PrivateTxStatus::Created,
+			creation_timestamp: self.mono_time.lock().to_system_time(),
+			validators: validator_logs,
+			deployment_timestamp: None,
+			public_tx_hash: None,
+		});
+	}
+
+	/// Logs the validation of the private transaction by one of its validators
+	pub fn signature_added(&self, tx_hash: &H256, validator: &Address) {
+		let mut logs = self.logs.write();
+		if let Some(transaction_log) = logs.get_mut(&tx_hash) {
+			if let Some(ref mut validator_log) = transaction_log.validators.iter_mut().find(|log| log.account == *validator) {
+				transaction_log.status = PrivateTxStatus::Validating;
+				validator_log.validation_timestamp = Some(self.mono_time.lock().to_system_time());
+			}
+		}
+	}
+
+	/// Logs the final deployment of the resulting public transaction
+	pub fn tx_deployed(&self, tx_hash: &H256, public_tx_hash: &H256) {
+		let mut logs = self.logs.write();
+		if let Some(log) = logs.get_mut(&tx_hash) {
+			log.status = PrivateTxStatus::Deployed;
+			log.deployment_timestamp = Some(self.mono_time.lock().to_system_time());
+			log.public_tx_hash = Some(*public_tx_hash);
+		}
+	}
+
+	fn read_logs(&self) -> Result<SystemTime, Error> {
+		let mut transaction_logs = self.logs_serializer.read_logs()?;
+		// Drop old logs
+		let earliest_possible = SystemTime::now().checked_sub(MAX_STORING_TIME).ok_or(Error::TimestampOverflow)?;
+		transaction_logs.retain(|tx_log| tx_log.creation_timestamp > earliest_possible);
+		// Sort logs by their creation time in order to find the most recent
+		transaction_logs.sort_by(|a, b| b.creation_timestamp.cmp(&a.creation_timestamp));
+		let initial_timestamp = transaction_logs.first()
```

Comment:
hrm, I don't understand the reasoning about returning max(l.creation_timestamp, SystemTime::now())

will not `SystemTIme::now() > l.creation_timestamp` always be true?

Why are we return `SystemTime::now` instead `most_recent` BTW?

Also, map and map_or could be merged to:

```.map_or(SystemTime::now(), |l| std::cmp::max(SystemTime::now(), l.creation_timestamp));```

Reply:
> will not SystemTIme::now() > l.creation_timestamp always be true?

No, `SystemTime` is not monotonic, see: https://doc.rust-lang.org/std/time/struct.SystemTime.html